### PR TITLE
Logging (1/2): new delay detector

### PR DESCRIPTION
**Pull request by nikurt**
Improves the logging system to automatically export durations of spans as metrics and detect spans taking too long.
This PR compliments the functionality Grafana Tempo.

Changes:
* Adds a tracing layer that keeps track of the duration of spans.
* Exports durations of spans to Prometheus.
* Logs spans that take longer than 1 sec to level=DEBUG target=delay_detector.

Performance impact should be minimal.
Memory usage increase is minimal.

### CODE REVIEW

File: core/o11y/src/delay_detector.rs [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+        // Keep the number of buckets small to limit the memory usage.
+        Some(exponential_buckets(0.01, 2.15443469, 10).unwrap()),
+    )
+    .unwrap()
+});
+
+struct Timings {
+    idle: u64,
+    busy: u64,
+    last: Instant,
```
Comment (wacban) [inline core/o11y/src/delay_detector.rs:31]: Can you add comments to those please?

File: core/o11y/src/delay_detector.rs [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+use prometheus::{exponential_buckets, HistogramVec};
+use std::time::Instant;
+use tracing::span::Attributes;
+use tracing::Id;
+use tracing_subscriber::layer::Context;
+use tracing_subscriber::registry::LookupSpan;
+use tracing_subscriber::Layer;
+
+#[derive(Default)]
+pub(crate) struct DelayDetectorLayer {}
```
Comment (wacban) [inline core/o11y/src/delay_detector.rs:12]: How about TracingSpanDurationLogger or Subscriber?

**Review by wacban — APPROVED** [review]
🥇

File: core/o11y/src/delay_detector.rs (hunk 2 above)
Comment (nikurt) [reply core/o11y/src/delay_detector.rs:12]: Thank you for the suggestion. `SpanDurationLogger` should be clear enough.

File: core/o11y/src/delay_detector.rs (hunk 1 above)
Comment (nikurt) [reply core/o11y/src/delay_detector.rs:31]: Done

File: core/o11y/src/span_duration_logger.rs [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+        let mut extensions = span.extensions_mut();
+        extensions.insert(Timings::new());
+    }
+
+    fn on_enter(&self, id: &Id, ctx: Context<'_, S>) {
+        let span = ctx.span(id).expect("Span not found, this is a bug");
+        let mut extensions = span.extensions_mut();
+        if let Some(timings) = extensions.get_mut::<Timings>() {
+            let now = Instant::now();
+            timings.idle += (now - timings.last).as_nanos() as u64;
```
[the remark this answers is not in the document]
Comment (nikurt) [reply core/o11y/src/span_duration_logger.rs:65]: Note that the `tracing-subscriber` crate handles durations as nanoseconds:
https://docs.rs/tracing-subscriber/latest/src/tracing_subscriber/fmt/fmt_layer.rs.html#1184

**Pull request merged**